node/node.go

```diff
@@ -2432,3 +2463,13 @@ func (p *proposalConsumerHare) OnProposal(proposal *types.Proposal) error {
 	}
 	return p.hare4.OnProposal(proposal)
 }
+
+type beaconGetter struct {
+	client *client.NodeService
+}
+
+func (b *beaconGetter) GetBeacon(e types.EpochID) (types.Beacon, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	return b.client.Beacon(ctx, e)
+}
```

Comment:
so the call eventually gets here (`beacon/beacon.go`):
```go
// GetBeacon returns the beacon for the specified epoch or an error if it doesn't exist.
func (pd *ProtocolDriver) GetBeacon(targetEpoch types.EpochID) (types.Beacon, error) {
	beacon := pd.getBeacon(targetEpoch)
	if beacon != types.EmptyBeacon {
		return beacon, nil
	}

	beacon, err := pd.getPersistedBeacon(targetEpoch)
	if err == nil {
		return beacon, nil
	}

	if errors.Is(err, sql.ErrNotFound) {
		return types.EmptyBeacon, errBeaconNotCalculated
	}
	pd.logger.Error("failed to get beacon from db",
		zap.Uint32("target_epoch", targetEpoch.Uint32()),
		zap.Error(err),
	)
	return types.EmptyBeacon, fmt.Errorf("get beacon from DB: %w", err)
}
```
Even if I do pass a context on the beacon interface, I'll have to add it everywhere where it is used, but then finally ignore it in this method.
For me the solution which I am suggesting is the smaller price to pay.

Reply:
That's one of the two implementations. The other one goes over the network. The fact that the first one doesn't support cancelation doesn't mean that the latter cannot. 

IMHO, it should be the _consumer_ (the entity that uses the interface) who defines the interface, not the _producer_ (the entity that implements it) (or a "global" interface like in this case). On the consumer side, it makes a lot of sense to pass in a context to allow cancelation.

> Even if I do pass a context on the beacon interface, I'll have to add it everywhere where it is used, but then finally ignore it in this method.

How about following the rule "define interfaces in the consumer package, not the producer" and define what's needed by the Oracle in the `eligibility` pkg:

```go
package eligibility

type BeaconProvider {
    Beacon(context.Context, types.EpochID) (types.Beacon, error)
}
```

and use it in place of `system.BeaconGetter`.

Then, if you want to avoid changing the implementation of `func (pd *ProtocolDriver) GetBeacon` in the `beacon` pkg and related "millions" of changes, you could write an adapter in the other direction:

```go
type beaconGetter struct {
	provider *beacon.ProtocolDriver
}

func (b *beaconGetter) Beacon(_ context.Context, e types.EpochID) (types.Beacon, error) {
	return b.provider.GetBeacon(e)
}
```

Reply:
yeah this makes sense. thanks for the feedback. implemented accordingly